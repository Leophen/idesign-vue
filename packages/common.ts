import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
