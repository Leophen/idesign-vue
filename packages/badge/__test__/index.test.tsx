import { Badge } from '../index';
import { render } from '../../common'

describe('Badge 组件测试', () => {
  function renderSup(badge) {
    const wrapper = render(badge);
    return wrapper.find('.i-badge__content');
  }

  it('count', () => {
    expect(renderSup(<Badge />).text()).toBe('0');
    expect(renderSup(<Badge count='new' />).text()).toBe('new');
  });

  it('max count', () => {
    expect(renderSup(<Badge count={100} />).text()).toBe('99+');
    expect(renderSup(<Badge count={10} maxCount={9} />).text()).toBe('9+');
  });

  it('color', () => {
    const wrapper = render(<Badge color="red" />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('dot', () => {
    expect(renderSup(<Badge dot />).classes()).toContain('i-badge--dot');
  })

  it('size', () => {
    expect(renderSup(<Badge size="small" />).classes()).toContain('i-badge__size-small');
  })

  it('shape', () => {
    expect(renderSup(<Badge shape="round" />).classes()).toContain('i-badge--round');
  })

  it('offset', () => {
    expect(renderSup(<Badge offset={[10, 10]} />).attributes('style')).toMatch('right: 10px; margin-top: 10px');
  })
});

