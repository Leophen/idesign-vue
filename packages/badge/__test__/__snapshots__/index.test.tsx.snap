// Vitest Snapshot v1

exports[`Badge 组件测试 > color 1`] = `
<div
  class="i-badge"
>
  
  
  <sup
    class="i-badge__content i-badge--circle"
    style="background: red;"
  >
    
    0
    
  </sup>
</div>
`;
